txpool/txpool_test.go

```diff
@@ -1,1312 +1,1898 @@
-	"fmt"
 	"math/big"
-	"strconv"
-	"sync"
 	"testing"
+	"time"
 
 	"github.com/0xPolygon/polygon-sdk/chain"
 	"github.com/0xPolygon/polygon-sdk/crypto"
 	"github.com/0xPolygon/polygon-sdk/helper/tests"
-	"github.com/0xPolygon/polygon-sdk/network"
 	"github.com/0xPolygon/polygon-sdk/txpool/proto"
 	"github.com/0xPolygon/polygon-sdk/types"
+	"github.com/golang/protobuf/ptypes/any"
 	"github.com/hashicorp/go-hclog"
 	"github.com/stretchr/testify/assert"
 )
 
-var forks = &chain.Forks{
-	Homestead: chain.NewFork(0),
-	Istanbul:  chain.NewFork(0),
-}
-
 const (
 	defaultPriceLimit uint64 = 1
 	defaultMaxSlots   uint64 = 4096
+	validGasLimit     uint64 = 10000000
 )
 
 var (
-	addr1 = types.Address{0x1}
-	addr2 = types.Address{0x2}
+	forks = &chain.Forks{
+		Homestead: chain.NewFork(0),
+		Istanbul:  chain.NewFork(0),
+	}
+
+	nilMetrics = NilMetrics()
 )
+
+// addresses used in tests
 var (
-	nilMetrics = NilMetrics()
+	addr1  = types.Address{0x1}
+	addr2  = types.Address{0x2}
+	addr3  = types.Address{0x3}
+	addr4  = types.Address{0x4}
+	addr5  = types.Address{0x5}
+	addr6  = types.Address{0x6}
+	addr7  = types.Address{0x7}
+	addr8  = types.Address{0x8}
+	addr9  = types.Address{0x9}
+	addr10 = types.Address{0x10}
 )
 
-type mockStore struct {
-}
+// returns a new valid tx of slots size with the given nonce
+func newTx(addr types.Address, nonce, slots uint64) *types.Transaction {
+	// base field should take 1 slot at least
+	size := txSlotSize * (slots - 1)
+	if size <= 0 {
+		size = 1
+	}
 
-func (m *mockStore) GetNonce(types.Hash, types.Address) uint64 {
-	return 0
-}
+	input := make([]byte, size)
+	if _, err := rand.Read(input); err != nil {
+		return nil
+	}
 
-func (m *mockStore) GetBlockByHash(types.Hash, bool) (*types.Block, bool) {
-	return nil, false
+	return &types.Transaction{
+		From:     addr,
+		Nonce:    nonce,
+		Value:    big.NewInt(1),
+		GasPrice: big.NewInt(0).SetUint64(defaultPriceLimit),
+		Gas:      validGasLimit,
+		Input:    input,
+	}
 }
 
-func (m *mockStore) GetBalance(types.Hash, types.Address) (*big.Int, error) {
-	balance, _ := big.NewInt(0).SetString("10000000000000000000", 10)
+// returns a new txpool with default test config
+func newTestPool() (*TxPool, error) {
+	return NewTxPool(
+		hclog.NewNullLogger(),
+		forks.At(0),
+		defaultMockStore{},
+		nil,
+		nil,
+		nilMetrics,
+		&Config{
+			PriceLimit: defaultPriceLimit,
+			MaxSlots:   defaultMaxSlots,
+			Sealing:    false,
+		},
+	)
+}
 
-	return balance, nil
+type accountState struct {
+	enqueued, promoted, nextNonce uint64
 }
 
-func (m *mockStore) Header() *types.Header {
-	return &types.Header{}
+type result struct {
+	accounts map[types.Address]accountState
+	slots    uint64
 }
 
-type mockSigner struct{}
+/* Single account cases (unit tests) */
+
+func TestAddTxErrors(t *testing.T) {
+	t.Run("ErrNegativeValue", func(t *testing.T) {
+		pool, err := newTestPool()
+		assert.NoError(t, err)
+
+		pool.EnableDev()
+		pool.SetSigner(crypto.NewEIP155Signer(100))
```

Comment:
There is already an issue created for pruning out this useless dev mode dependency from the TxPool, so we'll tackle it in some future sprint 👍 (next one)

I agree that @dbrajovic should just create a helper func inside the test for setting up the pool with the signer, since each test case uses the same block

Reply:
[Refactored](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/ee344186feb31d3e48a1ecba0c548ea69517fb2f)